include/wil/network.h

```diff
@@ -0,0 +1,1372 @@
+        {
+            return !(*this == rhs);
+        }
+
+        addr_info_iterator_t& operator++() WI_NOEXCEPT
+        {
+            WI_ASSERT(m_current_ptr);
+            if (m_current_ptr)
+            {
+                m_current_ptr = m_current_ptr->ai_next;
+
+                if (m_current_ptr)
+                {
+                    m_socket_address.reset(m_current_ptr->ai_addr, m_current_ptr->ai_addrlen);
+                }
+                else
+                {
+                    m_socket_address.reset();
+                }
+            }
+            else
+            {
+                m_socket_address.reset();
+            }
+            return *this;
+        }
+
+        addr_info_iterator_t operator++(int) WI_NOEXCEPT
+        {
+            auto return_value = *this;
+            this->operator++();
+            return return_value;
+        }
+
+        // begin and end are added here to enable support for range-based for loops
+        addr_info_iterator_t begin() const WI_NOEXCEPT
+        {
+            return *this;
+        }
+
+        constexpr addr_info_iterator_t end() const WI_NOEXCEPT
+        {
+            return {};
+        }
+
+    private:
+        // non-ownership of this pointer - the parent class must outlive the iterator
+        T* m_current_ptr{nullptr};
+        ::wil::network::socket_address m_socket_address{};
+    }; // class addr_info_iterator_t
+
+    //
+    // For example:
+    //     wil::unique_addrinfo resolved_addr;
+    //     if (0 == GetAddrInfoW(L"name_to_resolve.xyz", nullptr, nullptr, resolved_addr.addressof()))
+    //     {
+    //         for (const auto& address : wil::network::addr_info_iterator{resolved_addr.get()})
+    //         {
+    //             // address == wil::network::socket_address containing a resolved IP address
+    //         }
+    //     }
+    //
+    // If one can use exceptions:
+    //     const auto addresses = wil::network::resolve_name(name_to_resolve.xyz);
+    //     for (const auto& address : wil::network::addr_info_iterator{addresses.get()})
+    //     {
+    //         // address == wil::network::socket_address containing a resolved IP address
+    //     }
+
+    using addr_info_ansi_iterator = addr_info_iterator_t<ADDRINFOA>;
+    using addr_info_iterator = addr_info_iterator_t<ADDRINFOW>;
+    // not defining a type for ADDRINFOEXA as that type is formally __declspec(deprecated)
+    using addr_infoex_iterator = addr_info_iterator_t<ADDRINFOEXW>;
+
+#if defined(WIL_ENABLE_EXCEPTIONS)
+    // function to capture resolving IP addresses assigned to the local machine, throwing on error
+    // returning an RAII object containing the results
+    inline ::wil::unique_addrinfo resolve_name(PCWSTR name)
+    {
+        ADDRINFOW* addrResult{};
+        if (::GetAddrInfoW(name, nullptr, nullptr, &addrResult) != 0)
+        {
+            THROW_WIN32(::WSAGetLastError());

```

Comment:
One thing that might be worth considering is adding something like a `THROW_WSA_LAST_ERROR()` macro (and its friends). That would be useful to consumers outside of just this file.

Reply:
Not a bad idea. I looked at result_macros.h and how GetLastError is supported.
That ended up being a non-trivial # of edits to support that vast array of usage patterns.

So I think I'll defer this (good) idea to a later PR :)